src/coreclr/jit/flowgraph.cpp

```diff
@@ -711,6 +711,79 @@ bool Compiler::fgIsBlockCold(BasicBlock* blk)
     return ((blk->bbFlags & BBF_COLD) != 0);
 }
 
+//------------------------------------------------------------------------
+// fgIsSafeToRemoveCastOnAssignment: Determine whether "tree" is an assignment whose
+//                                   right-hand side is a cast that can be removed.
+//
+// Arguments:
+//    tree - The tree node under consideration
+//
+// Return Value:
+//    Returns true if the cast can be safely removed.
+//
+bool Compiler::fgIsSafeToRemoveCastOnAssignment(GenTree* tree)
+{
+    assert(tree->OperIs(GT_ASG));
+
+    GenTree* const op1 = tree->gtGetOp1();
+    GenTree* const op2 = tree->gtGetOp2();
+
+    GenTree* const effectiveOp1 = op1->gtEffectiveVal();
+
+    if (!effectiveOp1->OperIs(GT_IND, GT_LCL_VAR))
+        return false;
+
+    if (effectiveOp1->OperIs(GT_LCL_VAR) &&
+        !lvaGetDesc(effectiveOp1->AsLclVarCommon()->GetLclNum())->lvNormalizeOnLoad())
+        return false;
+
+    // If we are storing a small type, we might be able to omit a cast.
+    if (!varTypeIsSmall(effectiveOp1))
+        return false;
+
+    if (!op2->OperIs(GT_CAST))
+        return false;
+
+    if (op2->gtOverflow())
+        return false;
+
+    if (gtIsActiveCSE_Candidate(op2))
+        return false;
+
+    GenTreeCast* cast         = op2->AsCast();
+    var_types    castToType   = cast->CastToType();
+    var_types    castFromType = cast->CastFromType();
+
+    if (!varTypeIsIntegral(effectiveOp1))
+        return false;
+
+    if (!varTypeIsIntegral(castToType))
+        return false;
+
+    if (!varTypeIsIntegral(castFromType))
```

Comment:
Redundant with the `genActualType(castToType) != genActualType(castFromType)` check below.